clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -699,12 +706,19 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                 if (!parsed.isEmpty()) {
                     log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
                     parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
-                    ConsumerRecord<K, V> record = parsed.get(parsed.size() - 1);
-                    this.sensors.recordsFetchLag.record(partition.highWatermark - record.offset());
-                } else if (partition.highWatermark >= 0) {
-                    log.trace("Received empty fetch response for partition {} with offset {}", tp, position);
-                    this.sensors.recordsFetchLag.record(partition.highWatermark - fetchOffset);
                 }
+
+                if (partition.highWatermark >= 0) {
+                    log.trace("Received {} records in fetch response for partition {} with offset {}", parsed.size(), tp, position);
+                    Long partitionLag = subscriptions.partitionLag(tp);
+                    subscriptions.updateHighWatermark(tp, partition.highWatermark);
+                    if (partitionLag == null) {
```

Comment:
I added comments to explain this. This patch contains a bug fix to let the lag metric makes more sense. Previously the lag is the fetch lag, i.e. the lag of the fetched messages, but not consumed message. This is misleading because user may haven't seen the messages that has been fetched but not returned by `poll` yet. So this patch changed the metric value to be the lag based on the `position`. This logic is in `drainRecords()`. 

One problem is that the metric won't be updated in `drainRecords()` if the partition did not return any messages. So we are updating the lag the first time the consumer receives the fetch response from a partition here. If there are messages returned the lag will be updated again when the messages are returned to the user. And if there is no message returned, the lag will still be available. We only need to create the metric the first time the consumer receives the fetch response for a partition. That is why we check if the lag is null or not.